include/utilities/bmi/protocol.hpp

```diff
@@ -0,0 +1,208 @@
+/*
+Author: Nels Frazier
+Copyright (C) 2025 Lynker
+------------------------------------------------------------------------
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+------------------------------------------------------------------------
+
+Version 0.2
+Enumerate protocol error types and add ProtocolError exception class
+Implement error handling via expected<T, ProtocolError> and error_or_warning
+Removed model member and required model reference in run(), check_support(), and initialize()
+Minor refactoring and style changes
+
+Version 0.1
+Virtual interface for BMI protocols
+*/
+
+#pragma once
+
+#include "Bmi_Adapter.hpp"
+#include "JSONProperty.hpp"
+#include <nonstd/expected.hpp>
+
+namespace models{ namespace bmi{ namespace protocols{
+using nonstd::expected;
+using nonstd::make_unexpected;
+
+enum class Error{
+    UNITIALIZED_MODEL,
+    UNSUPPORTED_PROTOCOL,
+    INTEGRATION_ERROR,
+    PROTOCOL_ERROR,
+    PROTOCOL_WARNING
+};
+
+class ProtocolError: public std::exception {
```

Comment:
@PhilMiller, can you take a closer look at this class? Specifically, object construction / destruction and the `what()` method. This is getting into c++ territory I would like to phone an expert to look at :).